File: Algorithm/src/twopointers/C3_M_MiniWindowSort/C3_M_MiniWindowSort.java
package twopointers.C3_M_MiniWindowSort;

public class C3_M_MiniWindowSort {
  public static int miniWindow(int[] arr){
    int start = 0, end = arr.length - 1;

    //find the substring
    while(start < arr.length - 1 && arr[start] <= arr[start + 1]){
      start++;
    }
    if(start == arr.length - 1) return 0;
    while(end > 0 && arr[end] >= arr[end - 1]){
      end--;
    }
    if(end == 0) return arr.length;

    //find the max and min
    int max = Integer.MIN_VALUE, min = Integer.MAX_VALUE;
    for(int i = start; i <= end; i++){
      min = Math.min(min, arr[i]);
      max = Math.max(max, arr[i]);
    }

    //include all number out of substring
    while(start > 0 && arr[start - 1] < min){
      start--;
    }
    while(end < arr.length -1 && arr[end + 1] < max){
      end++;
    }

    return end - start + 1;
  }

  public static void main(String[] args) {
    System.out.println(miniWindow(new int[]{1,2,5,3,7,10,9,12}));
  }
}

File: Algorithm/bin/twopointers/C3_M_MiniWindowSort/C3_M_MiniWindowSort.py
import math

def miniWindow(arr):
  low, high = 0, len(arr) - 1
  while low < len(arr) - 1 and arr[low] <= arr[low + 1]:
    low += 1
  if low == len(arr):
    return 0
  while high > 0 and arr[high] >= arr[high - 1]:
    high -= 1
  if high == 0:
    return len(arr)
  
  minNum = math.inf
  maxNum = -math.inf
  for k in range(low, high + 1):
    minNum = min(minNum, arr[k])
    maxNum = max(maxNum, arr[k])
  
  while low > 0 and arr[low - 1] > minNum:
    low -= 1
  while high < len(arr) - 1 and arr[high + 1] < maxNum:
    high += 1

  return high - low + 1

def main():
  print(miniWindow([1,2,5,3,7,10,9,12]))

main()